### ISSUE TITLE
[API Proposal]: SearchValues<string>.Contains(ReadOnlySpan<char> span)

### ISSUE THREAD

**Issue by epeshk** [dotnet/runtime#104977]
### Background and motivation

The current `SearchValues<string>.Contains(string)` method forces an user to allocate and copy the `string` from `ReadOnlySpan<char>` to perform the search.

`StringSearchValuesAhoCorasick` and `AsciiStringSearchValuesTeddy...` already accept `ReadOnlySpan<char>` as a search argument, but it is not a public API.

`StringSearchValuesBase` uses `HashSet<string>`, but can search for `ReadOnlySpan<char> using `AlternateLookup` feature.

### API Proposal

```csharp
namespace System.Buffers;

public static class SearchValuesExtensions
{
  public static bool Contains(this SearchValues<string> searchValues, ReadOnlySpan<char> span);
}
```

### API Usage

```csharp
ReadOnlySpan<char> word = ...;
SearchValues s = SearchValues.Create(words);

bool found = s.Contains(word);
```

### Alternative Designs

```csharp
namespace System.Buffers;

public class SearchValues<T> where T : IEquatable<T>?, allows ref struct;

public static class SearchValues {
  public static SearchValues<ReadOnlySpan<char>> CreateSpanLookup(ReadOnlySpan<string> values, StringComparison comparisonType);
}
```

**Comment by stephentoub** [comment]
Can you share more information about where / how you want to use this? e.g. a more complete code example?

**Comment by jilles-sg** [comment]
Is this a documentation issue? The original poster may actually want `FrozenSet` which supports `GetAlternateLookup`. On the other hand, the reference page for `SearchValues<T>` at https://learn.microsoft.com/en-us/dotnet/api/system.buffers.searchvalues-1?view=net-9.0 does not clarify that the main use case of this type is with `IndexOfAny` and similar methods of `MemoryExtensions`, not the `Contains` instance method.

Alternatively, if the original poster has a set of strings and needs both searches for exact matches and entries of the set as substrings of the input string, the new API would make sense.

**Comment by epeshk** [comment]
@jilles-sg Yes, I'm definitely looking for the `MemoryExtensions.IndexOfAny(ReadOnlySpan<char>, SearchValues<string>)` method. I mistakenly confused it with the `Contains`.

For `Contains`, it's probably better to use hash set. To get benefits from `SearchValues`, it should be a very specific use case where it's worth avoiding hash code computation.

IMHO, it would be worth mentioning somewhere in the `SearchValues<T>` docs that it is intended to be used with the `MemoryExtensions` class.

**Comment by MihaZupan** [comment]
Docs were improved in https://github.com/dotnet/dotnet-api-docs/pull/12989
